File: java/src/main/java/banking/BankService.java
package banking;

import java.util.List;

public interface BankService {

    AccountStatus makeTransaction(Transaction transaction)
            throws ConnectionException,
            AccountNotFoundException,
            AccountBlockedException,
            NotEnoughFundsException;

    List<Transaction> listTransactions(String accountId)
            throws ConnectionException,
            AccountNotFoundException,
            AccountBlockedException;

    class ConnectionException extends Exception {
        public ConnectionException() {
            super("unable to connect to bank");
        }
    }

    class AccountNotFoundException extends Exception {
        public AccountNotFoundException() {
            super("account not found");
        }
    }

    class AccountBlockedException extends Exception {
        public AccountBlockedException() {
            super("account blocked");
        }
    }

    class NotEnoughFundsException extends Exception {
        public NotEnoughFundsException() {
            super("not enough funds");
        }
    }

}

File: kotlin/src/main/kotlin/banking/BankService.kt
package banking

interface BankService {

    @Throws(ConnectionException::class,
            AccountNotFoundException::class,
            AccountBlockedException::class,
            NotEnoughFundsException::class)
    fun makeTransaction(transaction: Transaction): AccountStatus

    @Throws(ConnectionException::class,
            AccountNotFoundException::class,
            AccountBlockedException::class)
    fun listTransactions(accountId: String): List<Transaction>

    class ConnectionException : RuntimeException("unable to connect to bank")
    class AccountNotFoundException : RuntimeException("account not found")
    class AccountBlockedException : RuntimeException("account blocked")
    class NotEnoughFundsException : RuntimeException("not enough funds")
}
